### ISSUE TITLE
Draw Calls even for off-screen and invisible objects.

### ISSUE THREAD

**Issue by tunepunk** [Scirra/Construct-3-bugs#1370]
## Problem description

Working with a big layout i noticed draw calls were unusually high even as some areas of the map didn't have a lot of objects. Seems like draw calls are issued even for invisible and off screen sprites (as long as they are moving, rotating or doing anything). In my mind this sounds like very wasteful and causing high draw call CPU usage in big layouts.  Maybe something that can be optimized as well for c3 runtime or a bug? objects are way off screen and invisible, on an invisible layer.

Edit: On further investigation draw calls are also issied for off screen non-moving objects if you use any type of scrolling. (which means everything in layout counts as "moving" and will generate draw calls)

## Attach a .c3p

https://www.dropbox.com/s/l1mw5g3nzmqlk4k/drawcalls.c3p?dl=1

## Steps to reproduce

1. Run attached project using debugger.
2. Notice high draw calls even for invisible off-screen objects.

## Observed result

Draw calls even for off screen sprites and invisible sprites. 

## Expected result

Hidden and off screen objects, and layers should not be drawn, and not generate draw calls.

## Affected browsers

* Chrome: yes
* Firefox: yes

**Comment by AshleyScirra** [comment]
Offscreen objects really aren't drawn. I can't see any performance difference here either (~1% draw calls either way).

Objects are at least iterated (with negligible performance overhead) to test if they're in the viewport, but this is not a factor unless you have 10,000+ instances - and then you can enable render cells to reduce the overhead.

Most likely I think is the fact the runtime actually stops rendering frames completely if nothing in the layout is changing at all. So I think you are benchmarking drawing at 60 FPS vs. not drawing at all.

**Comment by tunepunk** [comment]
Let me rephrase that and please download updated example file on same link. It's hard to explain what I'm experiencing but here are my results for 10000 objects instead. as i'm testing on a slower device. 

All are off screen and i'm just checking the draw calls.

10000 rotating off screen sprites with visibility set to on: 
Draw calls 22% CPU.

10000 rotating off screen sprites with visibility set to off: 
Draw calls 12% CPU.

10000 static off screen sprites with visibility set to on: 
Draw calls 26% CPU.

10000 static off screen sprites with visibility set to off: 
Draw calls 24% CPU.

When off screen sprites are set rotating but invisible, draw calls drop from 24% to 12-13% about half, compared to the other states. I'm sorry if my initial post didn't come out clear. Do you see there's a quite significant drop in draw calls cpu usage, by setting them rotating and invisible if they are off screen.

Which means the other three states use more CPU, and i have no clue why, but seems I'm saving a lot of CPU by rotating and setting invisible off screen stuff. It's very odd.....i know.

**Comment by tunepunk** [comment]
I find it really really odd. Why would the draw call iteration be twice as fast for sprites that are rotating and invisible?

and why would you need to iterate if invisible objects are on screen? if their visibility is set to hidden they don't need to be drawn anyway, so there's no point to check if they are on screen even. if there's a list of excluded uid's couldn't iteration be faster? and if you used the "is on screen" condition to set offscreen sprites invisible, they could be excluded from iterations too. Off screen stuff should probably be considered ivisible by default, as you can't see it. So wouldn't a custom SOL with only visable on screen objects be a lot faster than checking every single instance in layout if they are on screen?

maybe that's what scenario 2 is doing? no clue. half as much cpu for draw calls if the objects are set to moving and invisible. Boggles my mind.

**Comment by tunepunk** [comment]
And another thing. Draw calls CPU usage were exactly the same, regardless if they were on screen or not.

**Comment by tunepunk** [comment]
I redid the test on my Mac (chrome) at work as well, and here are the results.
https://www.dropbox.com/s/l1mw5g3nzmqlk4k/drawcalls.c3p?dl=1

10000 rotating off screen sprites with visibility set to on:
Draw calls 21% CPU. (I guess this on is highest because it has to check constantly updating bounding boxes, but when i used the "is on screen" condition in the project draw-calls dropped to 6.5% same as below test.) 

10000 rotating off screen sprites with visibility set to off:
Draw calls 2,5% CPU.

10000 static off screen sprites with visibility set to on:
Draw calls 6.5% CPU.

10000 static off screen sprites with visibility set to off:
Draw calls 2.5% CPU.

So by manually setting off-screen sprites to invisible can reduce the CPU load for draw calls, especially if you're already using the "is on screen" condition anywhere in the project.

I know this seems like small numbers, but as I'm targeting and testing on lower end devices as well all these numbers get blown up, as you have very low CPU power to deal with, and If i find something that could be optimized i try to find out what I can do about it, and what's causing it.

I'm just trying to make my projects run as good as possible even on lower end devices. Draw calls has always been one of the things bothering me, as it's always on the highest post, for any of my projects, and there's no way for me to optimize it with events, except for setting off screen objects to invisible. Setting off screen objects to invisible seems to exclude objects from bounding box checks to screen, and therefor, reduce the draw calls.

**Comment by AshleyScirra** [comment]
OK, if you've got something like 10,000 instances, then the overhead of iterating them to see if they should be drawn becomes significant. The "is visible" flag is checked before the "is in viewport" check, which is ever so slightly more work (four comparisons for each edge of the viewport rectangle instead of one). So if you flip the "is visible" flag off, it skips the viewport check; if you leave it visible, it will do the viewport check and then decide to not draw it. So basically you're profiling the different ways of deciding why not to draw an instance.

This is what render cells are for. It prevents even iterating far-away instances, much like collision cells avoid even checking far-away instances. Don't think you should set the visible flag off; that's the wrong solution, it will still iterate them and have some performance overhead. Turn on render cells and it doesn't even iterate them, so they have zero performance overhead.

**Comment by tunepunk** [comment]
Thanks for the feedback. Yes render cells is a great feature and I use them frequently on larger layouts. but for medium size layouts let's say (3x3 screen size or around that) i guess you're iterating (4/9 45%) of the content most of the time. If Cell size was half screen size, you would iterate max 9 out of 36 cells. (25% max), and 25 of 144 (17%) with 1/4 cell size.

![rendercells](https://user-images.githubusercontent.com/27011641/37338031-7a265552-26b6-11e8-8dab-cc529f651bf2.jpg)

I'm wondering if there would be any benefit of having a setting for the cell size? 1/2 or 1/4 of screen size, for some benefits on medium size layouts? 

Thanks for your reply. I'm just trying to squeeze out most out of the performance on slower devices.

**Comment by AshleyScirra** [comment]
It's a very difficult issue and depends on a lot of things. The visible cells have to have their contents merged and sorted which has a performance overhead for in-viewport content. In certain circumstances making the cell size smaller will reduce performance; in others, making the cell size larger will reduce performance. It depends on how much content is in the viewport vs. outside it. I think using the viewport size is a reasonable compromise. TBH if the layout is medium size, iterating every instance shouldn't be that much overhead - it's designed for very large layouts.
